### PR TITLE
POC implementation of attributes

### PR DESCRIPTION
**Pull request by kalmanb**
Mucking around with Model and Field. Feel free to use or delete.
Tested the UI, didn't run CI tests.

### CODE REVIEW

File: generated/app/BryzekApidocSpecV0Client.scala [hunk 1]
```diff
@@ -15,10 +15,26 @@ package com.bryzek.apidoc.spec.v0.models {
...
+   * necessarily define the API but aids in code generation. Examples would be hints
+   * for certain code generators about classes to extend, interfaces to implement,
+   * annotations to add, names to assign to certain methods, etc. The specific
+   * attributes will be applicable only in the context of the specific code
+   * generators usings them.
+   */
+  case class Attribute(
+    name: String,
+    value: _root_.play.api.libs.json.JsObject,
+    description: _root_.scala.Option[String] = None
```
Comment (mbryzek) [inline generated/app/BryzekApidocSpecV0Client.scala]: Seeing this now - we're going to want to add the deprecation object to the attribute for consistency.

**Comment by mbryzek** [comment]
This looks good!

Would you be interested in also taking a look at apidoc:api/app/lib/ServiceDiff.scala class? We'll want to include changes to attributes in the history of each service.

One piece that I'm not sure of how to implement is whether or not a change in an attribute should be considered a breaking change or not. My guess is that if we do not know, we should consider each attribute change a breaking change.

**Comment by kalmanb** [comment]
Cool, I'll add deprecation and have a look at ServiceDiff.

**Comment by kalmanb** [comment]
@mbryzek added deprecation and differences.

Breaking vs Nonbreaking is tricky. You could use attributes to change css styling in documentation and/or change critical validation. I made them non breaking for now. Happy to change. I did think that you could allow the user to define this as another field say 'breaking_change' on the attribute that specifies what to break on. Defaults to say "never"
```
{
  "name": "attribute1",
  "value": {},
  "breaking_change": ["never", "on_any", "on_addition", "on_removal", "on_value_change", "on_deprecation", "on_description_change"]
}
```

**Comment by mbryzek** [comment]
what did you think of the idea to require defining the attributes in the UI? would make it self documenting and would allow capture of the breaking change info up front

**Comment by kalmanb** [comment]
I didn't quite understand the use case, got it now. 
I like it. 
Would help generator developers specify available attributes. Could also make it easier to reuse common attributes and helps with typos. As you say fixes the breaking issue.
Would also clean up the apidoc.json api

Lot's to think about, are they at the org level, shareble, unique ID's, types, where can they be defined ...
Attribute:
* id (unique)
* element: model, field, etc
* description
* type: object, string, number, ...
* breaking: ...

Might be possible to specify them at the root level too:
```
"models": {
  "person": {
    "fields": ...
    "kalmanb.some_attr": "some text"
  }
}
```

**Comment by mbryzek** [comment]
maybe to start (and to keep scope in check) we make attributes globally unique. right now generators are also global - so I think it is logical to keep attributes the same

in the future if we decide to support private generators we can do same for attributes at that time

wdyt?

**Comment by kalmanb** [comment]
Yeah good call.
I've started using the above implementation. A big problem is when you make a mistake with the attribute you don't get any feedback when you upload the spec, only later from the generator. Having them native would allow validation when you upload the spec which is another win.

I'm happy to help out with the implementation. Let me know if you need a hand.

**Comment by mbryzek** [comment]
Awesome. Sounds like we have agreement and a design. Very happy for you to
implement - think what you've already done is great.

One small suggestion is that we create a separate pull request for the
postgresql schema for the attributes themselves - that way I can deploy the
schema to production in advance.

Is there any piece that you'd like me to help on? Otherwise I can review
the pull request and deploy to production.

**Comment by kalmanb** [comment]
Sounds good, I'll have a look into the UI and ping you if I get lost. 
Feel free to close off this PR.

**Comment by mbryzek** [comment]
Work is proceeding at PR #391

**Pull request closed without merging**